Connection-based DNS policy

raybejjani opened the pull request:
**Update 2019-11-04**
- A combo bug around how DNSCache.LookupByRegex returned names that didn't match (it returned an entry in the map with no IPs) caused cleanups to not happen when an IP finally expired out of zombies (and, previously, from the DNS caches). This is fixed in https://github.com/cilium/cilium/pull/9545. That commit is included here to simplify testing, but they should be merged separately, ideally.
- On startup the restore could use older versions of the cache, without GC happening (and the cascade to zombies). This is sorted out, although the code is a bit repetitive between the restore segment and dns-garbage-collector. I'm happy to refactor it, but it wasn't terribly clear how to do this while keeping the logic clear.

**Update 2019-11-01**
- renamed to DNSPendingDeletes DNSZombieMappings and switched the nomenclature to alive/dead instead of active/deletable
- DNSPendingDeletes.WantToDelete -> Upsert
  - DNSPendingDeletes.canDelete -> isAlive with the true/false sense switched around
 - Typo comment fixes
 - Added srcPort to matchCB/doFiltering in ctmap.gc
- Unit tests for DNSZombieMappings
- Reuse the set of Endpoints already obtained in the CT GC, avoiding repeated endpointmanager.LookupIP calls.


This PR adds deferring deleting IPs of allowed DNS names from the allowed list while a connection exists for that IP. Removing the policy will (I think) work as before, blocking the ongoing connection.
The general scheme here is to keep a list of to-delete IPs. When the CT GC runs, it marks active connections. If an IP is to-delete and marked active, it is not deleted. If a CT GC ends and an IP is not marked active, it is then deleted by the DNS GC controller `dns-garbage-collector-job`. 

Some things aren't working and I'll fix them before marking this pending:
- Policy doesn't regenerate on DNS expiration. This is a bug somewhere in how we propagate this event. I'll fix it in another PR. In the meantime, generating any other DNS lookup causes a regen.
- The delete lists need some cap. I might add that as a follow-up PR to limit the size of this one. Expect it to default to something large, like 10k IPs in the list, per endpoint. This means it would be a limit on ongoing connections for that particular endpoint once the DNS lookup for them has expired
- I need to change the minTTL defaults but I'll do this once this change is more baked in. The minimum is configurable today, so it's purely a sane defaults thing.

**`cilium fqdn cache list` output now shows**
```
Endpoint   Source       FQDN                   TTL   ExpirationTime             IPs
1865       lookup       flarg.cloudhmar.com.   27    2019-10-25T17:33:44.077Z   192.168.33.11
1865       connection   fleeg.cloudhmar.com.   0     2019-10-25T17:33:01.995Z   1.1.1.1
```

**Things to consider**
- The CT table entries are emitted by adding a callback to the GCFilter used. This is a hack. I originally passed a callback down the runGC callchain but it seemed more unwieldy overall.
- The CT scan can only mark active connections so deletions must happen indirectly to this loop.
- The Endpoint.DNSDeletes field should really live inside Endpoint.DNSHistory and the logic hidden there. This isn't possible because the JSON serialisation of DNSCache doesn't allow for new fields :/ (it is a list of the entries, and nothing else). Upgrades could work, but downgrades would not. Worse, these entries are immutable but the active-connection marking would make that untrue.
- We need to store the delete information on endpoints so that it is restored on startup. While we should be able to rebuild our state there is a race between the DNS GC controller, the CT GC goroutine, and the first regenerations of endpoints. Nonetheless, longer restart times might mean that DNS entries expire during the cilium-agent downtime. I've tried to account for this  (the first DNSHistory.GC call on each endpoint should do the right thing) but this may be racey.
- The refactoring of the cache expire logic to work in lockstep with DNSCache.GC calls means that the actual DNS expiration time has a minimum resolution of 1 minute. This is probably not an issue, but might cause confusion.


Fixes https://github.com/cilium/cilium/issues/9340

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/9497)
<!-- Reviewable:end -->

raybejjani committed:
ctmap: Add EmitCTEntryCB to GCFilter to expose per-entry data

We want to piggy-back on the CT table iteration. We already have
GCFilter.ValidIPs and .MatchIPs that are used to inject information.
EmitCTEntryCB exposes each entry allowed by GCFilter to be inspected
externally..

Signed-off-by: Ray Bejjani <ray@isovalent.com>

raybejjani committed:
api: /fqdn/cache responses include source field

The source field denotes the reason an entry is being returned. It can
be a lookup or a live connection based on an IP from a valid lookup.

Signed-off-by: Ray Bejjani <ray@isovalent.com>

raybejjani committed:
cmd: fqdn cache list shows the source of an entry

Signed-off-by: Ray Bejjani <ray@isovalent.com>

raybejjani committed:
fqdn: Add and use DNSZombieMappings in Endpoint

We need a way to delay expiring entries that are removed from DNSCache
instances in order to avoid blocking connections that exist longer than
the DNS TTLs.
DNSZombieMappings holds these expirations and allows each to be marked as
alive and in-use. This functionality should be held within DNSCache but
the serialization scheme precludes changing the structure (this would
work for upgrades but fail for downgrades).

Signed-off-by: Ray Bejjani <ray@isovalent.com>

raybejjani committed:
policy: Defer in-use DNS IP deletion

DNS TTLs reflect the validity of the DNS data but not of connections
made based on it. Our older expiration behaviour could cause existing
long-lived connections to be disconnected. The DNS GC controller now
accounts for this, delaying removal of IPs allowed by DNS and also used
by a connection. The connection information is extracted during the CT
(connection tracking) garbage collection.
The period of this GC varies, and may be as high as 12 hours for
LRUMaps. This interval is smaller when more connections are cleaned. In
that sense, there would be more DNS entries generated (and expired) when
there are more connections churning in the system. Thus, the lifetime of
a deferred IP deletion would be shorter since the CT GC happens more
often. In cases where there are few, but long-lived, connections the
deferred deletions would remain longer.

Signed-off-by: Ray Bejjani <ray@isovalent.com>

raybejjani committed:
fqdn: Tie DNSCache expiry to GC calls

We previously relied on natural TTL expiry to remove entries in the
cache. This included Lookup* functions eliding expired data when
returning. As we've added more varied expiration/eviction schemes to the
cache it has become more difficult to coordinate between caches. In
particular, it is difficult to be consistent between the
Endpoint.DNSHistory caches and the global cache, used by NameManager to
populate FQDNSelectors.
This change makes Lookup* functions use the latest cleanup time instead
of time.Now(). This relationship requires periodic .GC calls for the
cache in order to advance time and treat entries as expired.
To date, the only usage has been for toFQDNs and a controller is always
running to clear the caches.

Signed-off-by: Ray Bejjani <ray@isovalent.com>

raybejjani committed:
fqdn: DNSCache supports cascading to DNSZombieMappings

DNSCache can now cascade expirations and evictions to a
DNSZombieMappings instance, itself then responsible for final deletions
of toFQDNs allowed IPs. This leaves the handling of what happens to
expired entries to the user of DNSCache, allowing a choice between
retaining deleted entries or ignoring them.

Signed-off-by: Ray Bejjani <ray@isovalent.com>

joestringer commented on pkg/fqdn/cache.go:
```diff
@@ -233,50 +231,52 @@ func (c *DNSCache) addNameToCleanup(entry *cacheEntry) {
+// cleanupExpiredEntries cleans all the expired entries since lastCleanup up to
+// expires, but not including it. lastCleanup is set to expires and later
+// cleanups begin from that time.
+// It returns the list of names that have expired data.
```
nit:

```suggestion
// It returns the list of names that have expired data and a map of removed DNS cache entries, keyed by IP.
```

More elaborate types for the return values would be another way to document this, but I think a comment like the above would suffice.

joestringer commented on pkg/endpoint/endpoint.go:
```diff
@@ -190,6 +190,11 @@ type Endpoint struct {
+	// DNSDeletes is the collection of DNS IPs that have expired in or been
+	// evicted from DNSHistory. They are held back from deletion until we can
+	// confirm that no existing connection is using them.
+	DNSDeletes *fqdn.DNSPendingDeletes
```
Does anyone else get the urge to describe these entries as `ZombieDNSMappings`...?

raybejjani replied:
Yes! and it's almost halloween! I'll make that change first :P

joestringer commented on pkg/endpoint/fqdn.go:
```diff
@@ -0,0 +1,49 @@
+// dns-garbage-collector-job can correctly clear older connection data.
+func (e *Endpoint) MarkDNSCTEntry(dstIP net.IP, now time.Time) {
+	if dstIP == nil {
+		e.Logger(logSubsys).Error("MarkDNSCTEntry called with nil IP")
```
Should we include the IP in the log message?

raybejjani replied:
The endpoint IP? (the other one is nil). `e.Logger()` includes both IPs of the endpoint, along with a bunch of other stuff (quite a bit, in fact, these log lines must be pretty wide now!)

joestringer replied:
:man_facepalming: I read the error flow of this, came back, and read the log message and didn't put the two together. This is fine.

joestringer commented on pkg/endpoint/fqdn.go:
```diff
@@ -0,0 +1,49 @@
+// MarkDNSCTEntry records that dstIP is in use by a connection that is allowed
+// by toFQDNs policy. The reverse lookup is attempted in both DNSHistory and
+// DNSCTHistory, allowing short DNS TTLs but long-lived connections to
+// persisthere.DNSCTHistory is used to suppress delete handling of expired DNS
```
```suggestion
// persist there. DNSCTHistory is used to suppress delete handling of expired DNS
```

joestringer commented on pkg/endpoint/fqdn.go:
```diff
@@ -0,0 +1,49 @@
+// determine whether a new GC run has completed when comparing the timestamp
+// with the active timestamps of specific IPs marked with MarkDNSCTEntry.
+// NOTE: While the timestamp is ths start of the run, it should be set AFTER
+// the run completes to avoid races.
```
nit:
to avoid races between X and Y?

One of these would be the garbage collection controller, not sure what the other one is.

The comment above this function doesn't need to describe the races in detail, but it would be helpful to point out what is racing against what so that someone attempting to look at this can have an idea what they need to look at (or whether the other code is out-of-date compared to this comment).

raybejjani replied:
I might have included too much information, but I figured it was harmless.

joestringer commented on pkg/fqdn/cache.go:
```diff
@@ -644,3 +657,220 @@ func (c *DNSCache) UnmarshalJSON(raw []byte) error {
+
+// WantToDelete enqueues the ip -> qname as a possible deletion
+// updatedExisting is true when an earlier enqueue existed and was updated
+func (d *DNSPendingDeletes) WantToDelete(ipStr string, qname ...string) (updatedExisting bool) {
```
nit: This is usually named something like `Upsert()` ;)

raybejjani replied:
I resisted the urge to use `RaiseZombie` ;)

joestringer commented on pkg/fqdn/cache.go:
```diff
@@ -644,3 +657,220 @@ func (c *DNSCache) UnmarshalJSON(raw []byte) error {
+// expireLookupsBefore requires an entry to have been enqueued before the
+// specified time in order to remove it.
+// nameMatch will remove any DNS names that match.
+func (d *DNSPendingDeletes) ForceExpire(expireLookupsBefore time.Time, nameMatch *regexp.Regexp) (namesAffected []string) {
```
This looks like a reasonable candidate for new unit tests (generally pretty functional; has specific behaviour for empty vs. non-empty `time.Time`)

joestringer commented on pkg/fqdn/cache.go:
```diff
@@ -644,3 +657,220 @@ func (c *DNSCache) UnmarshalJSON(raw []byte) error {
+// nameMatcher (nil is match all) and when an entry no longer has any valid
+// names will it be removed outright.
+// Note that all parameters must match, if provided. `time.Time{}` is
+// considered not-provided for time parameters.
```
nit: The comment lost me at `considered not-provided`.

Would it be easier to provide a couple of examples?

```
// ForceExpire(time.Time{}, 'cilium.io') expires all entries for cilium.io.
// ForceExpire(time.Now(), 'cilium.io') expires all entries for cilium.io that expired before the current time.
```

raybejjani replied:
I had followed the other ForceExpire's comments. I've updated both to be more clear (I hope)

joestringer commented on pkg/fqdn/cache.go:
```diff
@@ -644,3 +657,220 @@ func (c *DNSCache) UnmarshalJSON(raw []byte) error {
+}
+
+// Delete after it is seen by CT GC, DeletePendingAt < lastCTGCUpdate, and the ActiveTime is 0 or no longer current, ActiveAt < lastCTGCUpdate
+func (d *DNSPendingDeletes) canDelete(del *DNSPendingDelete) bool {
```
nit: Looking at the way that the rest of this PR treats this value, I think it would be easier to follow if this was the opposite - either `isActive()` or `isAlive()`.

joestringer reviewed (commented):
Despite the various nits I commented on below that might help developers to
understand what's going on with this code, I was actually able to understand
what's going on here so you can take most of those suggestions with a grain of
salt. In colloquial conversation I will still call these pending deletions as
"zombies", whether you rename them in the code or not ;)

So if I follow right we already had a mark-and-sweep GC for DNS names which ran
once per minute. This PR changes this so that instead of marking and sweeping
to delete DNS entries, instead this outer mark-and-sweep removes the entries
from the global cache and creates the zombie DNS mappings associated with
individual endpoints that are using those DNS names.

With that change in place, there's now a second level of mark-and-sweep where
instead of running both mark and sweep from the `dns-garbage-collector-job`,
there's now split responsibilities of mark from the CT GC goroutine plus the
sweep tied into this same `dns-garbage-collector-job` (outer) sweep phase.
However for this second level rather than sweeping all entries since the
beginning of the current outer sweep time (running once per minute via
`dns-garbage-collector-job`), we use the start time of the most recent CT GC
run. This way we only clean entries that we know haven't seen active traffic
recently; it takes a double-tap to kill a zombie: once for DNS TTL expiry and
once for inactive connection confirmation.

I'd imagine that given the current code structure there's a reason that the two
mark-and-sweeps are both managed from the `dns-garbage-collector-job`
controller, and that you've considered splitting them. This might aid
understanding and allow better decoupling of the mechanisms, but I don't see a
particularly strong argument to go and rework it to work like that right now.

So my remaining questions are these:
* What kind of testing/validation will we do on this?
* If we assume that in principle, extending the lifetime of individual DNS
  entries as they apply for policy generation for endpoints is OK (which I
  think is the whole problem we're trying to solve here); what's the threshold
  for active connection validation?

For the latter question, the simple actionable item would be figuring out how
to bound CT GC run interval or at least consider DNS TTL validation as one
piece of feedback into the CT GC interval calculation: via sheer number of
outstanding zombies? Proportion of zombies out of all DNS names?

Longer term I could imagine us also (/alternatively?) hooking into connection
close events in the kernel on the local node and using those to trigger the
deletion of the corresponding DNS entries (or at least some kind of check). But
I expect there's a lot more infrastructure needed for that in Cilium.

aanm commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -215,8 +215,15 @@ type GCFilter struct {
 }
 
+// MatchCBFunc is the type used for the MatchCB callback in GCFilter
+type MatchCBFunc func(srcIP net.IP, dstIP net.IP, dstPort uint16, nextHdr, flags uint8, entry *CtEntry) bool
```
where's the srcPort? :thinking:

raybejjani replied:
It wasn't passed to doFiltering (which calls MatchCB). I thought it wasn't available but it is. I'll add it.

aanm commented on pkg/fqdn/cache.go:
```diff
@@ -233,23 +231,24 @@ func (c *DNSCache) addNameToCleanup(entry *cacheEntry) {
 func (c *DNSCache) cleanupExpiredEntries(expires time.Time) ([]string, time.Time) {
-	timediff := int(expires.Sub(c.lastCleanup).Seconds())
 	startTime := c.lastCleanup
+	if c.lastCleanup.Equal(time.Time{}) {
```
`c.lastCleanup.IsZero()`?

raybejjani replied:
+100!!!

aanm commented on pkg/fqdn/cache.go:
```diff
@@ -640,3 +640,220 @@ func (c *DNSCache) UnmarshalJSON(raw []byte) error {
+		d.deletes[ipStr] = entry
+	}
+
+	entry.Names = append(entry.Names, qname...)
```
Is it worth avoiding duplicated entries?

raybejjani replied:
Yes. I'll dedup them here

aanm commented on pkg/fqdn/cache.go:
```diff
@@ -640,3 +640,220 @@ func (c *DNSCache) UnmarshalJSON(raw []byte) error {
+}
+
+// Delete after it is seen by CT GC, DeletePendingAt < lastCTGCUpdate, and the ActiveTime is 0 or no longer current, ActiveAt < lastCTGCUpdate
+func (d *DNSPendingDeletes) canDelete(del *DNSPendingDelete) bool {
```
I'm having troubles understanding this function.
```
// Delete after it is seen by CT GC, DeletePendingAt < lastCTGCUpdate, and the ActiveTime is 0 or no longer current, ActiveAt < lastCTGCUpdate
```

> // Delete after it is seen by CT GC

which logic is doing this?

>  DeletePendingAt < lastCTGCUpdate

go it, is `d.lastCTGCUpdate.After(del.DeletePendingAt)`

> ActiveTime is 0 or no longer current

what's "current"?

> ActiveAt < lastCTGCUpdate

go it, is `d.lastCTGCUpdate.After(del.ActiveAt)`

Perhaps the part I'm having more trouble understanding is the field names. To me it looks like we can delete if the last CT GC update was executed after the time in the field "delete pending at" which is fine. Now for the second part, we can only delete if the last CT GC update was executed after the entry was active or 0? What does it mean an entry to be "active"?

raybejjani replied:
This function took me a while to get right the first time... and when I switched it to isAlive I ended up messing it up again!

One note on the new version of this (isAlive): there aren't any `time.BeforeOrEqual` type functions so I actually kept the older logic and added a not to the start :/

>>// Delete after it is seen by CT GC

> which logic is doing this?

The logic for the CT GC is in `pkg/maps/ctmap/gc/gc.go`. I can state that directly here if you want (but it is mentioned in the Endpoint function that calls this function).

>> ActiveTime is 0 or no longer current

> what's "current"?

"current" was a bad word. In the context of the CT GC cycle, current meant that something was active/alive during the most recent CT GC cycle. The comment doesn't use current anymore, but it may still not be clear.

Basically we can only delete an entry (now called a zombie) if a CT GC cycle completed and it was not marked active/alive. The code for this would look like `lastCTGCUpdate.After(DeletePendingAt) && lastCTGCUpdate.After(ActiveAt)` but, we need to guard against `ActiveAt` not being set, because only the CT GC can set it but the DNS garbage collector can check if entries/zombies are alive/active. Since DeletePendingAt is always set, 
In the new code, this function is called `isAlive` but it's actually easier to keep the canDelete logic and invert it. The alternative statement would be much larger, and we would need to check `ActiveAt.IsZero()`. It would be something like:
`(DeletePendingAt.After(lastGCUpdate) || DeletePendingAt.Equal(lastGCUpdate)) ||
  (AliveAt.After(lastGCUpdate) || AliveAt.Equal(lastGCUpdate))`

aanm commented on pkg/fqdn/cache.go:
```diff
@@ -640,3 +640,220 @@ func (c *DNSCache) UnmarshalJSON(raw []byte) error {
+
+		if !d.canDelete(cpy) {
+			active = append(active, cpy)
+			continue perIP
```
Why this jump?

raybejjani replied:
It was to skip adding to the deletable list just below. I changed it to be a simple if-else since that's what it is

aanm commented on pkg/fqdn/cache.go:
```diff
@@ -640,3 +640,220 @@ func (c *DNSCache) UnmarshalJSON(raw []byte) error {
+	for _, entry := range d.deletes {
+		// Do not expire entries that were enqueued after expireLookupsBefore, but
+		// only if the value is non-zero
+		if !expireLookupsBefore.Equal(time.Time{}) && entry.DeletePendingAt.After(expireLookupsBefore) {
```
`expireLookupsBefore.IsZero()`

aanm commented on pkg/fqdn/cache.go:
```diff
@@ -640,3 +640,220 @@ func (c *DNSCache) UnmarshalJSON(raw []byte) error {
+			continue
+		}
+
+		cpy := &DNSPendingDelete{
```
Why don't we setup a target to create a deepcopy method of this structure instead?

aanm commented on pkg/fqdn/cache.go:
```diff
@@ -640,3 +640,220 @@ func (c *DNSCache) UnmarshalJSON(raw []byte) error {
+	// Collect entries we can delete
+perIP:
+	for _, del := range d.deletes {
+		cpy := &DNSPendingDelete{
```
Why don't we setup a target to create a deepcopy method of this structure instead?

aanm commented on pkg/fqdn/cache.go:
```diff
@@ -640,3 +640,220 @@ func (c *DNSCache) UnmarshalJSON(raw []byte) error {
+// MarshalJSON encodes DNSPendingDeletes into JSON. Only the DNSPendingDelete
+// entries are encoded.
+func (d *DNSPendingDeletes) MarshalJSON() ([]byte, error) {
+	aux := struct {
```
add comment here stating we are doing this to avoid making `Deletes` as a public field for `DNSPendingDeletes`

aanm commented on pkg/maps/ctmap/gc/gc.go:
```diff
@@ -51,18 +51,47 @@ func Enable(ipv4, ipv6 bool, restoredEndpoints []*endpoint.Endpoint, mgr Endpoin
+					if flags != ctmap.TUPLE_F_OUT {
+						return true
+					}
+					if ep := mgr.LookupIP(srcIP); ep != nil {
```
This lookup will get an internal mutex for the manager for each CT entry. Can't we simply do a `mgr.GetEndpoints` before iterating CT map and do a lockless lookup map instead? We already have all endpoints on line `80` we just need to transform into a `map[ipStr]*Endpoint`

raybejjani replied:
I actually had it that way at the start of this thing... At the time it didn't make sense to keep it, but the way it works now it will be fine. I switched it back

joestringer commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -215,8 +215,15 @@ type GCFilter struct {
+
+	// MatchCB is called, when non-nil, if filtering by ValidIPs and MatchIPs
+	// passes. Returning true keeps the CT entry, and false deletes it.
+	MatchCB MatchCBFunc
```
nit:
```suggestion
	DeleteEntry DeleteEntryCB
```

or

```suggestion
	KeepEntry KeepEntryCB
```

(given that the boolean result of this function has an impact on whether connections are kept in the connection tracker or not)

Or alternatively, keep it as match and just don't return the boolean for deletion handling - I don't think that functionality is actually used anywhere right? So unless we have a plan for it to ever return false, why even support that for now?

raybejjani replied:
I felt bad adding it and calling it matchCB without it doing something "legit". I'll switch it to purely be there to emit the entry and drop the boolean return part. That will be more clear. (The match part was because the other fields were matches on IPs to delete or not but none of that is terribly clear here)

joestringer commented on daemon/fqdn.go:
```diff
@@ -152,35 +155,79 @@ func (d *Daemon) bootstrapFQDN(restoredEndpoints *endpointRestoreState, preCache
-			// cleanup caches for all existing endpoints as well.
+			var (
+				GCStart      = time.Now()
+				namesToClean []string
```
nit: The commit message for the second commit in the PR here still refers to `DNSDeletes`.

joestringer commented on pkg/fqdn/cache.go:
```diff
@@ -108,7 +111,10 @@ type DNSCache struct {
+	// LastCleanup is the latest time for which entries have been expired. It is
+	// used as "now" when doing lookups and advanced by calls to .GC
+	// When an entry is added with an expiration time before lastCleanup, it is
+	// set to that value.
```
Why can entry insertion affect the `DNSCache` most recent cleanup time? Is it something to do with how you would ensure to zombify the entry?

raybejjani replied:
It has to do with how (in this PR) cleanupTime is now used. Expiration is done by iterating over a map of expiration-time -> entries. The time is in seconds so it is somewhat grouped. `cleanupTime` historically limited the number of iterations (the older code derived this in the `timediff` variable) by setting the start of the iteration, and the `now` parameter set the end. This used to work fine because lookups also elided expired data so the external view of the cache was consistent in real-time (i.e. if the cleanup controller had not already executed, a lookup would already not include expired entries).
Now that I explicitly rely on `cleanupTime` a lookup may return expired entries (as intended) but I was worried that it would advance too far and never go back, thus never expiring that entry. I think it was also a way to set the value on first insert, but that has other solutions. I guess it is to help with zombification, in that this more explicit increment of lastCleanup ensures that the `DNSCache.GC()` calls return all expired entries for that iteration of the DNS garbage collector.

joestringer reviewed (approved):
Couple of other minor nits, but LGTM.

raybejjani committed:
fqdn: DNSCache LookupByRegex functions don't return empty matches

An accident in the logic that collected IPs that matched regexes
would always add a name, but correctly return no ips for it. This caused
other code to perceive the name as matching-with-ips. This subsequently
meant that names that had all names expired would not be correctly
cleaned in the policy map until another update required it.

DNSCache.LookupByRegex doesn't return names with no matching IP, and
mapIPToSelectors guards against this more robustly, matching the code
for .MatchName fields.

Signed-off-by: Ray Bejjani <ray@isovalent.com>
